lib/paperclip/storage/s3v2.rb

```diff
@@ -0,0 +1,451 @@
+          config[opt] = s3_credentials[opt] if s3_credentials[opt]
+        end
+
+        config.merge(@s3_options)
+      end
+
+      def obtain_s3_instance_for(options)
+        instances = (Thread.current[:paperclip_s3_instances] ||= {})
+        instances[options] ||= Aws::S3::Resource.new(options) # was: Aws::S3::Client.new(options)
+      end
+
+      def s3_bucket
+        @s3_bucket ||= s3_interface.bucket(bucket_name)
+      end
+
+      def s3_object style_name = default_style
+        s3_bucket.object(s3_object_name(style_name))
+      end
+
+      def s3_object_name style_name = default_style
+        path(style_name).sub(%r{\A/},'')
+      end
+
+      def using_http_proxy?
+        !!@http_proxy
+      end
+
+      def http_proxy_host
+        using_http_proxy? ? @http_proxy[:host] : nil
+      end
+
+      def http_proxy_port
+        using_http_proxy? ? @http_proxy[:port] : nil
+      end
+
+      def http_proxy_user
+        using_http_proxy? ? @http_proxy[:user] : nil
+      end
+
+      def http_proxy_password
+        using_http_proxy? ? @http_proxy[:password] : nil
+      end
+
+      def set_permissions permissions
+        permissions = { :default => permissions } unless permissions.respond_to?(:merge)
+        permissions.merge :default => (permissions[:default] || :public_read)
+      end
+
+      def set_storage_class(storage_class)
+        storage_class = {:default => storage_class} unless storage_class.respond_to?(:merge)
+        storage_class
+      end
+
+      def parse_credentials creds
+        creds = creds.respond_to?('call') ? creds.call(self) : creds
+        creds = find_credentials(creds).stringify_keys
+        env = Object.const_defined?(:Rails) ? Rails.env : nil
+        (creds[env] || creds).symbolize_keys
+      end
+
+      def exists?(style = default_style)
+        if original_filename
+          s3_interface.client.head_object(bucket: bucket_name, key:s3_object_name(style)).present?
+        else
+          false
+        end
+      rescue Aws::Errors::ServiceError
+        false
+      end
+
+      def s3_permissions(style = default_style)
+        s3_permissions = @s3_permissions[style] || @s3_permissions[:default]
+        s3_permissions = s3_permissions.call(self, style) if s3_permissions.respond_to?(:call)
+        s3_permissions
+      end
+
+      def s3_storage_class(style = default_style)
+        @s3_storage_class[style] || @s3_storage_class[:default]
+      end
+
+      def s3_protocol(style = default_style, with_colon = false)
+        protocol = @s3_protocol
+        protocol = protocol.call(style, self) if protocol.respond_to?(:call)
+
+        if with_colon && !protocol.empty?
+          "#{protocol}:"
+        else
+          protocol.to_s
+        end
+      end
+
+      def create_bucket
+        s3_interface.bucket(bucket_name).create
+      end
+
+      def flush_writes #:nodoc:
+        @queued_for_write.each do |style, file|
+        retries = 0
```

Comment:
Useless assignment to variable - `retries`.<br>Use 2 (not 0) spaces for indentation.